### PR TITLE
logging: impl custom log option per route

### PR DESCRIPTION
**Pull request by artur-barsegyan**

### CODE REVIEW

**Comment by sharonovd** [comment]
Generally LGTM for me

File: test/http.test.lua [hunk 1]
```diff
@@ -388,4 +413,81 @@ test:test("server requests", function(test)
...
+
+        httpd:stop()
+    end)
+
+    test:test("Log route requests with turned off 'log_requests' option", function(test)
+        test:plan(1)
+        local httpd_without_logs = http_server.new("127.0.0.1", 12345, { log_requests = false, log_errors = false })
+        httpd_without_logs:start()
+
+        httpd:route({ path = '/', log_level = 3 }, function(_) end)
```
Comment (ASverdlov) [inline test/http.test.lua]: `httpd_without_logs:route` maybe?
After this fix the test should fail.

File: README.md [hunk 2]
```diff
@@ -158,6 +158,10 @@ The first argument for `route()` is a Lua table with one or more keys:
 * `path` - route path, as described earlier.
 * `name` - route name.
 * `method` - method on the route like `POST`, `GET`, `PUT`, `DELETE`
+* `log_level` - type of logger specified for logging request info.
+  Any digit from range 1 - 6 is allowed, where: `1` is disable logging - `6` is `log.verbose`.
+  By default, `log_level` is equal to 5 (`log.info`). If `log_requests` option is `false` this parameter will be ignored.
```
Comment (ASverdlov) [inline README.md]: > If `log_requests` option is `false` this parameter will be ignored. 
This is not true. According to your code, `log_level` just rewrites a log function and doesn't depend on `log_requests` value anyhow.

File: test/http.test.lua (hunk 1 above)
Comment (artur-barsegyan) [reply test/http.test.lua]: yeah
you right. copy - paste is evil

File: http/server.lua [hunk 3]
```diff
@@ -589,7 +589,23 @@ response_mt = {
...
+    return log.info
+end
+
+local function set_error_logger(route)
+    if route and route.endpoint.err_logger ~= nil then
+        return route.endpoint.err_logger
+    end
+
+    return log.error
+end
```
Comment (filonenko-mikhail) [inline http/server.lua]: функции больше похожи на get_*** чем на set_***

Comment (artur-barsegyan) [reply http/server.lua]: in english, please!
soglasen Mish

**Comment by ASverdlov** [comment]
We have another feature that is really needed: custom logger.
If we accept this feature, it won't solve it and will introduce unneeded complexity in the future.

I propose adding new route options (which we have in server already): `log_errors `, `log_requests`. Extend their semantic in both server and route: accept `function` type (with log.info semantics) and `boolean` type (legacy behaviour).

In case route option is specified - it overrides the server one.

**Comment by ASverdlov** [comment]
As discussed privately, LuaJit errors on sockets introduced by new tests need to be fixed.

**Comment by artur-barsegyan** [comment]
> As discussed privately, LuaJit errors on sockets introduced by new tests need to be fixed.

I post a new comment to [this bug issue ](https://github.com/tarantool/tarantool/issues/4087). Actually, it's tarantool socket implementation bug

File: README.md [hunk 4]
```diff
@@ -110,8 +110,13 @@ httpd = require('http.server').new(host, port[, { options } ])
   type `text/html`, `text/plain` and `application/json`.
 * `display_errors` - return application errors and backtraces to the client
   (like PHP).
-* `log_errors` - log application errors using `log.error()`.
-* `log_requests` - log incoming requests.
+* `log_requests` - log incoming requests. This parameter can receive:
+    - custom log function using for all routes in current instance of http server. The function should receive a format string
```
Comment (ASverdlov) [inline README.md]: Faster to read version:

function value, supporting C-style formatting: `log_requests(fmt, ...)`, where `fmt` is a format string and `...` is Lua Varargs, holding arguments to be replaced in fmt.

File: README.md [hunk 5]
```diff
@@ -110,8 +110,13 @@ httpd = require('http.server').new(host, port[, { options } ])
...
   (like PHP).
-* `log_errors` - log application errors using `log.error()`.
-* `log_requests` - log incoming requests.
+* `log_requests` - log incoming requests. This parameter can receive:
+    - custom log function using for all routes in current instance of http server. The function should receive a format string 
+      as the first argument and the following variable number of arguments uses with format string.
+    - boolean value, where `true` choose default `log.info` and `false` disable request logs at all. 
+  
+  By default uses `log.info` function for requests logging. 
+* `log_errors` - same as the `log_requests` option but uses for error messages logging. By default uses `log.error()` function.
```
Comment (ASverdlov) [inline README.md]: Grammar:
> but uses for
but is used for

File: http/server.lua [hunk 6]
```diff
@@ -589,7 +589,53 @@ response_mt = {
     }
 }
 
-local function handler(self, request)
+local function is_function(obj)
+    return type(obj) == 'function'
+end
+
+local function get_request_logger(server_opts, route_opts)
+    local root_logger = log.debug
```
Comment (ASverdlov) [inline http/server.lua]: Simplification suggestion:
1) check route first, if found - return immediately, then
2) check server, if found - return immediately, then
3) return default

No need for `root_logger` variable.

File: http/server.lua [hunk 7]
```diff
@@ -589,7 +589,53 @@ response_mt = {
...
+        elseif route_opts.endpoint.log_requests == false then
+            return log.debug
+        end
+    end
+
+    return root_logger
+end
+
+local function get_error_logger(server_opts, route_opts)
+    local root_logger = log.debug
```
Comment (ASverdlov) [inline http/server.lua]: Simplification suggestion:
1) check route first, if found - return immediately, then
2) check server, if found - return immediately, then
3) return default

No need for `root_logger` variable.

File: http/server.lua [hunk 8]
```diff
@@ -685,21 +729,22 @@ local function process_client(self, s, peer)
             s:write('HTTP/1.0 100 Continue\r\n\r\n')
         end
 
-        local logreq = self.options.log_requests and log.info or log.debug
+        local route = self:match(p.method, p.path)
```
Comment (ASverdlov) [inline http/server.lua]: Add comment describing what you did here!

Something like:
`-- HACK: we're in a server context, but prematurely dispatch route (router responsibility) for getting logging options which override server ones' `.

File: http/server.lua [hunk 9]
```diff
@@ -685,21 +729,22 @@ local function process_client(self, s, peer)
...
+        local logreq = get_request_logger(self.options, route)
         logreq("%s %s%s", p.method, p.path,
-            p.query ~= "" and "?"..p.query or "")
+               p.query ~= "" and "?"..p.query or "")
 
-        local res, reason = pcall(self.options.handler, self, p)
+        local res, reason = pcall(self.options.handler, self, route, p)
```
Comment (ASverdlov) [inline http/server.lua]: Maybe better leave double-dispatch and not change router signature?

File: http/server.lua (hunk 8 above)
Comment (artur-barsegyan) [reply http/server.lua]: I think this is obvious and no comment is needed.

File: http/server.lua (hunk 9 above)
Comment (artur-barsegyan) [reply http/server.lua]: Yes, it will be more obvious

**Comment by ASverdlov** [comment]
LGTM.

**Pull request merged**